File: pkg-r/tests/testthat/test-retired.R
# Version retirement: remote OLS obsolete terms and cache retired sidecars.

test_that(".normalize_obo turns short forms and IRIs into colon obo_ids", {
  expect_identical(.normalize_obo("GO_0006915"), "GO:0006915")
  expect_identical(
    .normalize_obo("http://purl.obolibrary.org/obo/MONDO_0005016"),
    "MONDO:0005016"
  )
  expect_null(.normalize_obo(NULL))
  expect_null(.normalize_obo(""))
})

# An OLS transport stub that serves one obsolete term with the given successor.
.stub_obsolete <- function(replaced_by) {
  function(url, timeout) {
    list(
      status = 200,
      body = list(
        page = list(totalElements = 1),
        "_embedded" = list(
          terms = list(list(
            is_obsolete = TRUE,
            term_replaced_by = replaced_by
          ))
        )
      )
    )
  }
}

test_that("remote mode marks an obsolete term invalid with its successor", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_obsolete("GO_0006915"))

  res <- check_id("GO:0006917", source_db = "go", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$normalized))
  expect_identical(res$suggestion, "GO:0006915")
})

test_that("remote mode surfaces a cross-ontology successor from a full IRI", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = .stub_obsolete(
      "http://purl.obolibrary.org/obo/MONDO_0005016"
    )
  )

  res <- check_id("EFO:0000401", source_db = "efo", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$normalized))
  expect_identical(res$suggestion, "MONDO:0005016")
})

test_that("cache mode marks a retired id invalid with its successor", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())

  res <- check_id(
    "GO:0006917",
    source_db = "go",
    how = "cache",
    version = "sample"
  )
  expect_false(res$valid)
  expect_true(is.na(res$normalized))
  expect_identical(res$suggestion, "GO:0006915")
})

test_that("a well-formed id neither in the snapshot nor retired has no suggestion", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())

  res <- check_id(
    "GO:0000001",
    source_db = "go",
    how = "cache",
    version = "sample"
  )
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that(".snapshot_retired reads the bundled go retired sidecar", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())

  retired <- .snapshot_retired("go", "sample")
  expect_true("GO:0006917" %in% names(retired))
  expect_identical(retired[["GO:0006917"]], "GO:0006915")
})

test_that(".snapshot_retired returns an empty named vector with no sidecar", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())

  retired <- .snapshot_retired("mondo", "sample")
  expect_identical(length(retired), 0L)
  expect_identical(names(retired), character(0))
})

File: pkg-py/tests/test_retired.py
"""Version-retirement support: remote obsolete terms and cache retired maps."""

import pytest

import biogate
import biogate._remote as remote
from biogate._cache import _snapshot_retired
from biogate._remote import _normalize_obo


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOGATE_CACHE_DIR", str(tmp_path))


def test_normalize_obo_short_form():
    assert _normalize_obo("GO_0006915") == "GO:0006915"


def test_normalize_obo_full_iri():
    iri = "http://purl.obolibrary.org/obo/MONDO_0005016"
    assert _normalize_obo(iri) == "MONDO:0005016"


def test_normalize_obo_empty_and_none():
    assert _normalize_obo(None) is None
    assert _normalize_obo("") is None


def _stub_obsolete(ident, term_replaced_by):
    """Return a _http_get stub serving one obsolete OLS term for ``ident``."""

    def _http_get(url, timeout=30):
        if f"obo_id={ident}" in url:
            return 200, {
                "page": {"totalElements": 1},
                "_embedded": {
                    "terms": [
                        {"is_obsolete": True, "term_replaced_by": term_replaced_by}
                    ]
                },
            }
        return 404, None

    return _http_get


def test_remote_obsolete_suggests_successor(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub_obsolete("GO:0006917", "GO_0006915"))
    res = biogate.check_id("GO:0006917", source_db="go", how="remote")[0]
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion == "GO:0006915"


def test_remote_obsolete_cross_ontology_successor(monkeypatch):
    iri = "http://purl.obolibrary.org/obo/MONDO_0005016"
    monkeypatch.setattr(remote, "_http_get", _stub_obsolete("EFO:0000401", iri))
    res = biogate.check_id("EFO:0000401", source_db="efo", how="remote")[0]
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion == "MONDO:0005016"


def test_cache_retired_suggests_successor():
    res = biogate.check_id("GO:0006917", source_db="go", how="cache", version="sample")[
        0
    ]
    assert res.valid is False
    assert res.normalized is None
    assert res.suggestion == "GO:0006915"


def test_cache_not_retired_has_no_suggestion():
    res = biogate.check_id("GO:0000001", source_db="go", how="cache", version="sample")[
        0
    ]
    assert res.valid is False
    assert res.suggestion is None


def test_snapshot_retired_map_for_go_sample():
    assert _snapshot_retired("go", "sample") == {"GO:0006917": "GO:0006915"}
